hr_contract_cae/models/hr_contract.py

```diff
@@ -1,12 +1,137 @@
 from dateutil.relativedelta import relativedelta
 
-from odoo import api, fields, models
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ContractGroup(models.Model):
+    _name = "hr.contract.group"
+    _description = "Contract Group"
+    _order = "amendment_index"
+
+    contract_ids = fields.One2many(
+        comodel_name="hr.contract",
+        inverse_name="contract_group_id",
+        string="Contracts",
+    )
+
+    def get_initial(self):
+        return self.contract_ids.search(
+            [("contract_group_id", "=", self.id)], limit=1
+        )
+
+    def get_parent(self, amendment_index):
+        return self.contract_ids.search(
+            [
+                ("contract_group_id", "=", self.id),
+                ("amendment_index", "<", amendment_index),
+            ],
+            order="amendment_index desc",
+            limit=1,
+        )
+
+    def get_child(self, amendment_index):
+        return self.contract_ids.search(
+            [
+                ("contract_group_id", "=", self.id),
+                ("amendment_index", ">", amendment_index),
+            ],
+            limit=1,
+        )
+
+    def get_latest(self):
+        return self.contract_ids.search(
+            [("contract_group_id", "=", self.id)],
+            order="amendment_index desc",
+            limit=1,
+        )
+
+
+class ContractType(models.Model):
+
+    _inherit = "hr.contract.type"
+
+    echelon = fields.Selection(
+        [("main", "Main"), ("amendment", "Amendment")],
+        string="Echelon",
+        default="main",
+    )
+    max_usage = fields.Integer(string="Maximum Usage", default=False)
 
 
 class Contract(models.Model):
     _inherit = "hr.contract"
 
-    reason = fields.Char(string="Reason for recourse to contract")
+    name = fields.Char(compute="_compute_name")
```

Comment:
You should store the name because
- it is very regularly accessed
  - besides avoiding recomputation, it will (i'm 75% sure) allow caching the name
- it will be useful to have the name in the database